apps/keyboard/test/unit/imes/latin/worker_test.js

```diff
@@ -479,233 +477,258 @@ suite('Latin worker', function() {
           assert.equal(suggestions[i], expected[i], 'index: ' + i);
         }
       }
-
-      next();
-    });
+    }).then();
   }
 
-  function setupLanguage(langCode, keymap, userDictBlob, next) {
-    var loader = new InputMethodDatabaseLoader();
-    loader.start();
-    loader.SOURCE_DIR = '/js/imes/';
-    loader.load('latin', 'dictionaries/' + langCode + '.dict')
-    .then(function(dictData) {
-      worker.postMessage({
-        cmd: 'setLanguage',
-        args: [langCode, dictData, userDictBlob]
-      }, [dictData]);
-    })['catch'](function(e) { // workaround gjlsint error
-      console.error(e.toString());
-    });
 
-    worker.postMessage({
-      cmd: 'setNearbyKeys',
-      args: [keymap]
-    });
+  function setupLanguage(langCode, keymap, userDictBlob) {
+    return new Promise(function(resolve,reject){
+      var loader = new InputMethodDatabaseLoader();
+      loader.start();
+      loader.SOURCE_DIR = '/js/imes/';
+      loader.load('latin', 'dictionaries/' + langCode + '.dict')
+      .then(function(dictData) {
```

Comment:
I am sorry that I did not spot this on the previous review.

This function is a little bit complex because we want to do two `postMessage()` and wait for 4 `successCount`.

Since `loader.load('latin', 'dictionaries/' + langCode + '.dict')` also returns a promise, we should also attempt to move it into the chain and not to `catch()` it's error by ourselves. This should therefore be written as:

```javascript

  function setupLanguage(langCode, keymap, userDictBlob) {
    var loader = new InputMethodDatabaseLoader();
    loader.start();
    loader.SOURCE_DIR = '/js/imes/';
    var loaderPromise = 
      loader.load('latin', 'dictionaries/' + langCode + '.dict')
      .then(function(dictData) {
        worker.postMessage({
          cmd: 'setLanguage',
          args: [langCode, dictData, userDictBlob]
        }, [dictData]);
      });

    worker.postMessage({
      cmd: 'setNearbyKeys',
      args: [keymap]
    });

    var successCount = 0;
    var waitForSuccessPromise = new Promise(function(resolve, reject) {
      worker.onmessage = function(e) {
        if (e.data.cmd !== 'success') {
          reject('worker.onmessage unexpected result ' + e.data.message + '\n');
        }

        if (e.data.fn.startsWith('setLanguage') ||
            e.data.fn.startsWith('setNearbyKeys')) {
          successCount++;

          if (successCount === 4) {
            worker.onmessage = null;
            resolve();
          }
        }
      };
    });

    return loaderPromise
      .then(function() {
        return waitForSuccessPromise;
      });
  }

```

In this function, we chain the two promises and return the resulting promise. The returned promise therefore resolves when both promises are resolved.

I've also removed an `assert` from the `var waitForSuccessPromise = new Promise(function(resolve, reject) {` block because exception throw in that function will not cause promise to reject -- it would simply never call `resolve()`.